rfcs/proposed/task_group_dynamic_dependencies/implementation_details.md

```diff
@@ -0,0 +1,729 @@
+of successors.
+
+## Transferring successors to the other task
+
+The API `tbb::task_group::current_task::transfer_successors_to(tbb::task_handle& new_task)` works by setting `m_new_dynamic_state` in
+the dynamic state of the currently executing task to point to the dynamic state of `new_task`.
+
+To determine the dynamic state associated with the currently executing task, a pointer to the task must be obtained from the task scheduler.
+To achieve this, the new entry point `tbb::task* r1::current_task()` is added. 
+
+The implementation of `transfer_successors_to` is shown below:
+
+```cpp
+class task_group {
+    struct current_task {
+        static void transfer_successors_to(tbb::task_handle& new_task) {
+            tbb::task* curr_task = r1::current_task();
+            __TBB_ASSERT(curr_task != nullptr); // transfer called outside of the task
+            task_handle_task* curr_th_task = dynamic_cast<task_handle_task*>(curr_task);
+                // no assert(curr_th_task != nullptr) to allow no-op transfer_successors_to from function_stack_task
+            if (curr_th_task != nullptr) {
+                curr_th_task->transfer_successors_to(new_task->get_dynamic_state());
+            }
+        }
+    };
+};
+```
+
+`task_handle_task::transfer_successors_to` calls `task_dynamic_state::transfer_successors_to(new_task_state)` if the dynamic state of the
+current task is initialized. If the state is not initialized, there cannot be any successors associated with the current task.
+
+`task_dynamic_state::transfer_successors_to(task_dynamic_state* new_dynamic_state)` sets the `m_new_dynamic_state` pointer to `new_dynamic_state`,
+retrieves the successor list, and appends it to `new_dynamic_state`:
+
+```cpp
+class task_dynamic_state {
+private:
+    std::atomic<successor_list_node*> m_successor_list_head;
+    std::atomic<task_dynamic_state*>   m_new_dynamic_state;
+public:
+    void transfer_successors_to(task_dynamic_state* new_dynamic_state) {
+        // register current dynamic state as a co-owner of the new state
+        // to prolong its lifetime if the recipient task completes
+        // see dynamic state lifetime issue section for more details
+        new_dynamic_state->reserve();
+
+        m_new_dynamic_state.store(new_dynamic_state);
+        successor_list_node* list = m_successor_list_head.exchange(~std::uintptr_t(0));
+        new_dynamic_state->add_successor_list(list);
+    }
+};
+```
+
+`add_successor_list` is implemented similarly to `add_successor`, but without additional checks, since the recipient task
+can only be in the `created` state and cannot change.
+
+```cpp
+void task_dynamic_state::add_successor_list(successor_list_node* list) {
+    if (list != nullptr) {
+        // Finding the last node in the list
+        successor_list_node* last_node = list;
+
+        while (last_node->get_next() != nullptr) last_node = last_node->get_next();
```

Comment:
Is this traversal a potential performance concern, or do we not expect the lists to be big enough for real troubles?

We could in principle add a pointer to the tail of the list to the dynamic state. If extra space is a concern, it could be put into a union with `m_new_dynamic_state`.